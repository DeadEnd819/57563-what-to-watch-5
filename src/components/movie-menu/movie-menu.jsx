import React, {useCallback} from "react";
import {Link} from "react-router-dom";
import {MovieMenuType} from "../../prop-types/prop-types";
import {AppRoute} from "../../const";
import PlayButton from "../play-button/play-button";
import AddMyListButton from "../add-my-list-button/add-my-list-button";
import PropTypes from "prop-types";
import {fetchCurrentFilm, fetchReviews, updateFavoriteFilm} from "../../store/api-actions";
import {redirectToRoute} from "../../store/action";
import {connect} from "react-redux";

const MovieMenu = ({isMain, film, redirectAction, isUserLogged, loadDataFilm, addToMyListAction}) => {
  const {FILMS, REVIEW} = AppRoute;
  const {id, name, genre, released, isFavorite} = film;

  const handleAddMyListButtonClick = useCallback(
      () => {
        if (!isUserLogged) {
          redirectAction();
        }

        addToMyListAction(id, !isFavorite);
        loadDataFilm(id);
      }, [film, id, isFavorite]
  );

  return (
    <div className="movie-card__desc">
      <h2 className="movie-card__title">{name}</h2>
      <p className="movie-card__meta">
        <span className="movie-card__genre">{genre}</span>
        <span className="movie-card__year">{released}</span>
      </p>

      <div className="movie-card__buttons">
        <PlayButton id={id} />
        <AddMyListButton isFavorite={isFavorite} onButtonClick={handleAddMyListButtonClick} />
        {!isMain && <Link to={`${FILMS}/${id}${REVIEW}`} className="btn movie-card__button">Add review</Link>}
      </div>
    </div>
  );
};

MovieMenu.defaultProps = {
  isUserLogged: false,
};

MovieMenu.propTypes = {
  isMain: PropTypes.bool,
  film: MovieMenuType.isRequired,
  isUserLogged: PropTypes.bool.isRequired,
  loadDataFilm: PropTypes.func.isRequired,
  redirectAction: PropTypes.func.isRequired,
  addToMyListAction: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  loadDataFilm(id) {
    dispatch(fetchCurrentFilm(id));
    dispatch(fetchReviews(id));
  },
  addToMyListAction(id, isFavorite) {
    dispatch(updateFavoriteFilm(id, isFavorite));
  },
  redirectAction() {
    dispatch(redirectToRoute(AppRoute.LOGIN));
  },
});

export {MovieMenu};
export default connect(null, mapDispatchToProps)(MovieMenu);


