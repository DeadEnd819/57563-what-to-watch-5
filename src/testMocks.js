export const promoFilm = {
  id: 1,
  name: `The Grand Budapest Hotel`,
  posterImage: `the-grand-budapest-hotel-poster.jpg`,
  backgroundImage: `bg-the-grand-budapest-hotel.jpg`,
  backgroundColor: `#ffffff`,
  genre: `Comedy`,
  released: 2014,
  isFavorite: false,
};

export const film = {
  id: 1,
  name: `The Grand Budapest Hotel`,
  posterImage: `the-grand-budapest-hotel-poster.jpg`,
  previewImage: `bg-the-grand-budapest-hotel.jpg`,
  backgroundImage: `bg-the-grand-budapest-hotel.jpg`,
  backgroundColor: `#ffffff`,
  previewVideoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.
  Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
  rating: 8.9,
  scoresCount: 240,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`, `Saoirse Ronan`],
  runTime: 99,
  genre: `Comedy`,
  released: 2014,
  isFavorite: true,
};

export const reviewOne = {
  id: 1,
  user: {
    id: 4,
    name: `Kate Muir`,
  },
  rating: 8.9,
  comment: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
  date: `2019-05-08T14:13:56.569Z`,
};

export const reviewTwo = {
  id: 2,
  user: {
    id: 5,
    name: `Paula Fleri-Soler`,
  },
  rating: 9.1,
  comment: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
  date: `2019-08-08T17:40:56.569Z`,
};

export const reviews = [reviewOne, reviewTwo];

export const genresList = [
  `Action`,
  `Adventure`,
  `Comedy`,
  `Crime`,
  `Drama`,
  `Fantasy`,
  `Thriller`
];

export const state = {
  DATA: {
    films: [film, film, film, film, film, film, film],
    promo: promoFilm,
    currentFilm: film,
    currentReviews: reviews,
  },
  FILTER: {
    genre: `Drama`,
    showFilmsCount: 8,
  },
  USER: {
    name: `name`,
    avatar: `avatar.jpg`,
    status: `NOT_AUTHORIZED`,
    reviewStatus: `REVIEW_NOT_UPDATED`,
  },
};
