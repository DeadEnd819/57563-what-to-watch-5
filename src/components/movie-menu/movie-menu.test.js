import React from "react";
import ShallowRenderer from "react-test-renderer/shallow";
import {MovieMenu} from "./movie-menu";
import {film} from "../../testMocks";

jest.mock(`../play-button/play-button.jsx`, () => `PlayButton`);
jest.mock(`../add-my-list-button/add-my-list-button.jsx`, () => `AddMyListButton`);

const renderer = new ShallowRenderer();
const noop = ()=>{};

describe(`Render MovieMenu`, () => {
  it(`Should MovieMenu isMain render correctly`, () => {
    renderer.render(
        <MovieMenu
          film={film}
          isMain={true}
          isUserLogged={true}
          redirectAction={noop}
          loadDataFilm={noop}
          addToMyListAction={noop}
        />
    );

    const tree = renderer.getRenderOutput();

    expect(tree).toMatchSnapshot();
  });

  it(`Should MovieMenu render correctly`, () => {
    renderer.render(
        <MovieMenu
          film={film}
          isMain={false}
          isUserLogged={false}
          redirectAction={noop}
          loadDataFilm={noop}
          addToMyListAction={noop}
        />
    );

    const tree = renderer.getRenderOutput();

    expect(tree).toMatchSnapshot();
  });
});
