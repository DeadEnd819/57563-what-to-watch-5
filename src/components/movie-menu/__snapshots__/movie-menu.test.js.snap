// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render MovieMenu Should MovieMenu isMain render correctly 1`] = `
<div
  className="movie-card__desc"
>
  <h2
    className="movie-card__title"
  >
    The Grand Budapest Hotel
  </h2>
  <p
    className="movie-card__meta"
  >
    <span
      className="movie-card__genre"
    >
      Comedy
    </span>
    <span
      className="movie-card__year"
    >
      2014
    </span>
  </p>
  <div
    className="movie-card__buttons"
  >
    <PlayButton
      id={1}
    />
    <AddMyListButton
      isFavorite={true}
      onButtonClick={[Function]}
    />
  </div>
</div>
`;

exports[`Render MovieMenu Should MovieMenu render correctly 1`] = `
<div
  className="movie-card__desc"
>
  <h2
    className="movie-card__title"
  >
    The Grand Budapest Hotel
  </h2>
  <p
    className="movie-card__meta"
  >
    <span
      className="movie-card__genre"
    >
      Comedy
    </span>
    <span
      className="movie-card__year"
    >
      2014
    </span>
  </p>
  <div
    className="movie-card__buttons"
  >
    <PlayButton
      id={1}
    />
    <AddMyListButton
      isFavorite={true}
      onButtonClick={[Function]}
    />
    <Link
      className="btn movie-card__button"
      to="/films/1/review"
    >
      Add review
    </Link>
  </div>
</div>
`;
